src/lib/wallet/WalletConnectClient.js

```diff
@@ -0,0 +1,568 @@
+//@flow
+import { useCallback, useContext, useEffect, useState } from 'react'
+import WalletConnect from '@walletconnect/client'
+import web3Utils from 'web3-utils'
+import abiDecoder from 'abi-decoder'
+import Web3 from 'web3'
+import { first, sortBy } from 'lodash'
+import AsyncStorage from '../utils/asyncStorage'
+import { delay } from '../utils/async'
+import api from '../../lib/API/api'
+import logger from '../logger/js-logger'
+import { useSessionApproveModal } from '../../components/walletconnect/WalletConnectModals'
+import { GlobalTogglesContext } from '../contexts/togglesContext'
+import { useWallet } from './GoodWalletProvider'
+
+const log = logger.child({ from: 'WalletConnectClient' })
+
+//TODO:
+//7. cancel tx
+//8. edit gas
+//9. advanced edit tx values/contract call values
+//10. events
+//11. show warning if unable to decode contract call
+/**
+ * Parses the read WalletConnet URI from QR Code.
+ * If not valid, returns null.
+ * If valid, returns the WalletConnet URI.
+ * @param {string} link - receive WalletConnect URI
+ * @returns {string|null} - {link|null}
+ */
+export const readWalletConnectUri = link => {
+  // checks that the link has the expected strings in it
+  const eip1328UriFormat = /wc:[\w\d-]+@\d+\?bridge=.*&key=[a-z0-9]+/
+  const validUri = first(link.match(eip1328UriFormat))
+
+  if (!validUri) {
+    return null
+  }
+
+  return link
+}
+
+export const getWalletConnectTopic = link => {
+  const eip1328UriFormat = /wc:([\w\d-]+)@\d+\?bridge=.*&key=[a-z0-9]+/
+  const topic = link.match(eip1328UriFormat)[1]
+
+  return topic
+}
+
+const CHAINS_CACHE_KEY = 'chains'
+
+export const useChainsList = () => {
+  const { cache, hasCache, setCache } = useContext(GlobalTogglesContext)
+  const [chains, setChains] = useState(cache[CHAINS_CACHE_KEY])
+
+  useEffect(() => {
+    if (hasCache(CHAINS_CACHE_KEY)) {
+      return
+    }
+
+    api.getChains().then(chains => {
+      const ordered = sortBy(chains, 'name')
+
+      setChains(ordered)
+      setCache(CHAINS_CACHE_KEY, ordered)
+    })
+  }, [setChains, setCache, hasCache])
+
+  return chains
+}
+
+const cachedWeb3 = {}
+const getWeb3 = rpc => {
+  const web3 = cachedWeb3[rpc]
+  if (web3) {
+    return web3
+  }
+
+  const tempWeb3 = new Web3(new Web3.providers.HttpProvider(rpc))
+
+  cachedWeb3[rpc] = tempWeb3
+  return tempWeb3
+}
+
+const getChainRpc = chainDetails => {
+  return first((chainDetails.rpc || chainDetails.rpcUrls).filter(_ => _.includes('${') === false))
+}
+
+// Create connector
+let cachedConnector
```

Comment:
@sirpy global vars are evil.
any data should be encapsulated somewhere
for some global data React provides contexts, it should be used instead
ideally - a dedicated context for wallet connect stuff
but for now i've introduced a simple cache inside GlobalTogglesContext.

please store connector and web3 and other cache from global vars there
or add a separate ctx

Reply:
@sirpy as continuation of our midnight discussion

i see state-pool as the good compromise.
it satisfies the following criteria:

- it's tested and supported by its developers. so there's some guarantee credit. which means less probability of bugs / issues we may have in future
- it has the store encapsulating data. it gives more control and provides more data safety
- it's easier to test. exporting **let** is a huge violation rather than store object exported. with store object we could set any fixture for any unit tests we may need
- it's very simple, simplier than let-solution. you shouldn't perform an assignment and define local state var. you're getting an ready-to-use state var from the hook, cached globally under cover

here's an example of state-pool implementation: https://github.com/GoodDollar/GoodDAPP/pull/3802/commits/0e4b8a7d1f291b3d3809a6d8048944cbcc32c679

just compare my code with yours and see how it's simpler and easier

i definitely suggest you to go with this compromise and fix other let-implementations the same way

P.S. i also checked the whole official React docs and found nothing regarding module-wide let globals usage to keep data in the hooks. So the article you sent isn't an official position of the React team. anyway i agree for state-pool, but not the 'manual' let-solution in each particular case